src/annotator/integrations/image-text-layer.js

```diff
@@ -80,95 +208,107 @@ export class ImageTextLayer {
+          wordEl.style.display = 'inline-block';
+          wordEl.style.transformOrigin = 'top left';
+          wordEl.textContent = word.text;
+
+          if (prevWord) {
+            wordEl.style.marginLeft = scaledValue(
+              'x',
+              word.rect.left - prevWord.rect.right
+            );
+          }
+          prevWord = word;
+
+          // Set the size of this box used for layout. This does not affect the
+          // rendered size of the content.
+          wordEl.style.width = scaledValue('x', word.rect.width);
+          wordEl.style.height = scaledValue('y', word.rect.height);
+
+          // Don't collapse whitespace at end of words, so it remains visible
+          // in selected text. Also prevent line breaks due to overflows.
+          wordEl.style.whiteSpace = 'pre';
+
+          // Scale content using a transform. This affects the rendered size
+          // of the text, used by text selection and
+          // `Element.getBoundingClientRect`, but not layout.
+          const metrics = context.measureText(word.text);
+          const xScale = scaledValue('x', word.rect.width / metrics.width, '');
+          const yScale = scaledValue('y', word.rect.height / fontSize, '');
+          wordEl.style.transform = `scale(${xScale}, ${yScale})`;
+
+          lineEl.append(wordEl);
+        }
+
+        columnEl.append(lineEl);
       }
 
-      boxes.push({
-        span,
-        box: currentWordBox,
-        width: context.measureText(currentWordText).width,
-        height: boxHeight,
-      });
-
-      currentWordBox = null;
-      currentWordText = '';
-    };
-
-    for (let i = 0; i < charBoxes.length; i++) {
-      const char = text[i];
-      if (/\s/.test(char)) {
-        addCurrentWordToTextLayer();
-        continue;
-      }
-
-      const charBox = charBoxes[i];
-      currentWordBox = mergeBoxes(currentWordBox, charBox);
-      currentWordText += char;
+      container.append(columnEl);
     }
-    addCurrentWordToTextLayer();
 
-    // Position and scale text boxes to fit current image size.
-    const updateBoxSizes = () => {
+    const updateTextLayerSize = () => {
       const { width: imageWidth, height: imageHeight } =
         image.getBoundingClientRect();
       container.style.width = imageWidth + 'px';
       container.style.height = imageHeight + 'px';
 
-      for (let { span, box, width, height } of boxes) {
-        const left = box.left * imageWidth;
-        const top = box.top * imageHeight;
-        const right = box.right * imageWidth;
-        const bottom = box.bottom * imageHeight;
-
-        span.style.left = left + 'px';
-        span.style.top = top + 'px';
-
-        const scaleX = (right - left) / width;
-        const scaleY = (bottom - top) / height;
-
-        span.style.transform = `scale(${scaleX}, ${scaleY})`;
-      }
+      container.style.setProperty('--x-scale', `${imageWidth}`);
+      container.style.setProperty('--y-scale', `${imageHeight}`);
```

Comment:
Late-breaking thought here. I wonder if these custom property names could be more descriptive for debugging and clarity purposes.

Reply:
I've kept this as is for the moment. Open to review it in future.